### PR TITLE
[Roformer] Fixing a bug in RoFormerEncoder where it was ignoring the length of past_key_values when generating as a decoder

### PR DESCRIPTION
**Pull request by TheWall9**
# What does this PR do?

This pull request fixes the bug relating to the generation.

When using `RoFormerForCausalLM`, the `RoFormerEncoder` generates position embeddings based on the `seq_len` of the `hidden_states`. However, when using the `generate()` method to generate text token by token, the `seq_len` is always 1 and the length of `past_key_values` is ignored.

Using the solution from https://github.com/JunnYu/RoFormer_pytorch/blob/roformer_v2/src/roformer/modeling_roformer.py

My setup:
>- `transformers` version: 4.28.0.dev0
>- Platform: Windows-10-10.0.19045-SP0
>- Python version: 3.10.10
>- Huggingface_hub version: 0.13.3
>- PyTorch version (GPU?): 2.0.0 (False)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@JunnYu 
@ArthurZucker
@younesbelkada

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for fixing, this looks good to me, can you also run the styling checks?
I also left one open question

File: tests/models/roformer/test_modeling_roformer.py [hunk 1]
```diff
@@ -220,6 +220,30 @@ def create_and_check_for_causal_lm(
...
+        config.is_decoder = True
+        config.bos_token_id = 0
+        config.eos_token_id = 1
+        config.pad_token_id = 2
+        model = RoFormerForCausalLM(config=config)
+        model.to(torch_device)
+        model.eval()
+        with torch.no_grad():
+            input_ids = model.generate(do_sample=False, max_length=10)
+            output = model(input_ids=input_ids)
```
[the remark this answers is not in the document]
Comment (TheWall9) [reply tests/models/roformer/test_modeling_roformer.py:243]: When using the `generate()` method to generate text token by token, a unique sequence is obtained when using greedy decoding (`do_sample=False`). If the complete sequence is input to the model (`output = model(input_ids=input_ids)`), the resulting output should be consistent with the output generated by feeding the input text token by token, except for the `bos_token_id` added by `generate()` at the beginning and the last token generated by the complete sequence.

[the remark this answers is not in the document]
Comment (younesbelkada) [reply tests/models/roformer/test_modeling_roformer.py:243]: Thanks for clarifying! Can you run the styling checks? `make fixup`

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply tests/models/roformer/test_modeling_roformer.py:243]: Actually there is a test in transformers that ensures that the output with/without past are the same. I am wondering why it was not triggered. `test_attention_outputs` should be doing this. 
Otherwise we usually have this test implemented as: (example with T5)
```python 
    def create_and_check_generate_with_past_key_values(
        self,
        config,
        input_ids,
        decoder_input_ids,
        attention_mask,
        decoder_attention_mask,
        lm_labels,
    ):
        model = T5ForConditionalGeneration(config=config).to(torch_device).eval()
        torch.manual_seed(0)
        output_without_past_cache = model.generate(
            input_ids[:1], num_beams=2, max_length=5, do_sample=True, use_cache=False
        )
        torch.manual_seed(0)
        output_with_past_cache = model.generate(input_ids[:1], num_beams=2, max_length=5, do_sample=True)
        self.parent.assertTrue(torch.all(output_with_past_cache == output_without_past_cache))
```

[the remark this answers is not in the document]
Comment (TheWall9) [reply tests/models/roformer/test_modeling_roformer.py:243]: > Thanks for clarifying! Can you run the styling checks? `make fixup`

Thank you, `make fixup` version has been submitted

[the remark this answers is not in the document]
Comment (TheWall9) [reply tests/models/roformer/test_modeling_roformer.py:243]: > Actually there is a test in transformers that ensures that the output with/without past are the same. I am wondering why it was not triggered. `test_attention_outputs` should be doing this. Otherwise we usually have this test implemented as: (example with T5)
> 
> ```python
>     def create_and_check_generate_with_past_key_values(
>         self,
>         config,
>         input_ids,
>         decoder_input_ids,
>         attention_mask,
>         decoder_attention_mask,
>         lm_labels,
>     ):
>         model = T5ForConditionalGeneration(config=config).to(torch_device).eval()
>         torch.manual_seed(0)
>         output_without_past_cache = model.generate(
>             input_ids[:1], num_beams=2, max_length=5, do_sample=True, use_cache=False
>         )
>         torch.manual_seed(0)
>         output_with_past_cache = model.generate(input_ids[:1], num_beams=2, max_length=5, do_sample=True)
>         self.parent.assertTrue(torch.all(output_with_past_cache == output_without_past_cache))
> ```

Maybe the pre-training model released by roformer is mainly mask language model, test case does not take causal language model into account

[the remark this answers is not in the document]
Comment (younesbelkada) [reply tests/models/roformer/test_modeling_roformer.py:243]: After reading @ArthurZucker 's comment, I think it makes sense to add a test similar to what he suggested, what do you think about replacing the test you have implemented with the one suggested by Arthur?

[the remark this answers is not in the document]
Comment (TheWall9) [reply tests/models/roformer/test_modeling_roformer.py:243]: Not recommended by the author, but I wrote it based on the test cases of gpt2, which uses the predefined `expected_ output_ Ids`, but due to the lack of a suitable `RoFormerForCausalLM` pre training model, it is difficult to define `expected_ output_ Ids`, so I use the sequence generated by `generate()`

```
@require_torch
class GPT2ModelLanguageGenerationTest(unittest.TestCase):
    def _test_lm_generate_gpt2_helper(
        self,
        gradient_checkpointing=False,
        reorder_and_upcast_attn=False,
        scale_attn_by_inverse_layer_idx=False,
        verify_outputs=True,
    ):
        model = GPT2LMHeadModel.from_pretrained(
            "gpt2",
            reorder_and_upcast_attn=reorder_and_upcast_attn,
            scale_attn_by_inverse_layer_idx=scale_attn_by_inverse_layer_idx,
        )
        if gradient_checkpointing:
            model.gradient_checkpointing_enable()
        else:
            model.gradient_checkpointing_disable()
        model.to(torch_device)

        # The dog
        input_ids = torch.tensor([[464, 3290]], dtype=torch.long, device=torch_device)

        # The dog was found in a field near the intersection of West and West Streets.\n\nThe dog
        # fmt: off
        expected_output_ids = [
            464, 3290, 373, 1043, 287, 257, 2214, 1474, 262, 16246, 286, 2688, 290, 2688, 27262, 13, 198, 198, 464, 3290,
        ]
        # fmt: on
        output_ids = model.generate(input_ids, do_sample=False)
        if verify_outputs:
            self.assertListEqual(output_ids[0].tolist(), expected_output_ids)
```

**Review by younesbelkada — APPROVED** [review]
Awesome work! Thanks a lot for fixing

**Review by sgugger — APPROVED** [review]
Thanks for the fix! @ArthurZucker I'll let you merge if you're happy with this PR.

File: src/transformers/models/roformer/modeling_roformer.py [hunk 2]
```diff
@@ -262,6 +262,15 @@ def forward(
...
             value_layer = self.transpose_for_scores(self.value(hidden_states))
+            if sinusoidal_pos is not None:
+                if self.rotary_value:
+                    query_layer, key_layer, value_layer = self.apply_rotary_position_embeddings(
+                        sinusoidal_pos, query_layer, key_layer, value_layer
+                    )
+                else:
+                    query_layer, key_layer = self.apply_rotary_position_embeddings(
+                        sinusoidal_pos, query_layer, key_layer
+                    )
```
Comment (ArthurZucker) [inline src/transformers/models/roformer/modeling_roformer.py:273]: This is done twice (here and in the next else. This is a small nits!

**Review by ArthurZucker — APPROVED** [review]
There's a really small nits which allows de-duplicating code, otherwise LGTM! Thanks for addressing this

File: src/transformers/models/roformer/modeling_roformer.py (hunk 2 above)
Comment (TheWall9) [reply src/transformers/models/roformer/modeling_roformer.py]: Thank you for your carefulness, this is a great first PR experience for me.
